src/installer/managed/Microsoft.NET.HostModel/MachO/CodeSignature.cs

```diff
@@ -0,0 +1,155 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.IO.MemoryMappedFiles;
+using System.Linq;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.NET.HostModel.MachO;
+
+internal class CodeSignature
+{
+    private (EmbeddedSignatureHeader Header, long FileOffset) _embeddedSignature;
+    private (CodeDirectoryHeader Header, long FileOffset) _codeDirectory;
+    private (byte[] Identifier, long FileOffset) _identifierPtr;
+    private (byte[] Blob, long FileOffset) _cdHashes;
+    private (RequirementsBlob Header, long FileOffset) _requirementsBlob;
+    private (CmsWrapperBlob Header, long FileOffset) _cmsWrapperBlob;
+    private bool _unrecognizedFormat;
+
+    public uint Size => _embeddedSignature.Header.Size;
+
+    private CodeSignature() { }
+
+    public static bool AreEquivalent(CodeSignature a, CodeSignature b)
+    {
+        if (a is null ^ b is null)
+            return false;
+        if (a is null && b is null)
+            return true;
+        if (a._unrecognizedFormat || b._unrecognizedFormat)
+            return false;
+        if (!a._embeddedSignature.Equals(b._embeddedSignature))
+            return false;
+        if (!a._codeDirectory.Equals(b._codeDirectory))
+            return false;
+        if (!a._identifierPtr.Identifier.SequenceEqual(b._identifierPtr.Identifier))
+            return false;
+
+        var aSpecialSlotHashes = a._cdHashes.Blob.AsSpan(0, (int)MachObjectFile.SpecialSlotCount * MachObjectFile.DefaultHashSize);
+        var bSpecialSlotHashes = b._cdHashes.Blob.AsSpan(0, (int)MachObjectFile.SpecialSlotCount * MachObjectFile.DefaultHashSize);
+        if (!aSpecialSlotHashes.SequenceEqual(bSpecialSlotHashes))
+            return false;
+        var aCodeHashes = a._cdHashes.Blob.AsSpan(((int)MachObjectFile.SpecialSlotCount + 1) * MachObjectFile.DefaultHashSize);
+        var bCodeHashes = b._cdHashes.Blob.AsSpan(((int)MachObjectFile.SpecialSlotCount + 1) * MachObjectFile.DefaultHashSize);
+        if (!aCodeHashes.SequenceEqual(bCodeHashes))
+            return false;
+
+        return true;
+    }
+
+    public static CodeSignature Create(
+        (EmbeddedSignatureHeader Header, long FileOffset) embeddedSignature,
```

Comment:
Do we want duplicate FileOffsets at all? Since the signature blob looks contiguous it seems like we need only one file-relative offset -- the offset to the signature blob. Everything else looks like a relative offset.

Reply:
Yeah, technically we don't need the other pointers, I can remove them.